[Spring 지하철 노선도 - 1,2단계] 렉스(오성원) 미션 제출합니다.

Seongwon97 opened the pull request:
안녕하세요 코니👋🏻
레벨1 로또 미션 이후로 다시 뵙게 되었네요!!

이번에도 코니의 리뷰를 통해 많은 성장을 해보겠습니다!!

미션을 진행하며 생긴 궁금증들을 코드 코멘트로 남겼습니다🥸

감사합니다🙇🏻‍♂️

Seongwon97 committed:
docs: 요구사항 정리

Seongwon97 committed:
feat: 중복된 이름의 지하철역 예외 처리 구현

Seongwon97 committed:
feat: 컨트롤러와 서비스 레이어 분리 & 역 삭제 기능 구현

Seongwon97 committed:
feat: 의존성 관리 클래스 구현

Seongwon97 committed:
refactor: stationDao의 검증 로직을 service로 이동

Seongwon97 committed:
feat(LineDao): 지하철 노선 등록 기능 구현

Seongwon97 committed:
feat(LineDao): 지하철 노선 목록 조회 기능 구현

Seongwon97 committed:
feat(LineDao): 지하철 노선 조회 기능 구현

Seongwon97 committed:
feat(LineDao): 지하철 노선 수정 기능 구현

Seongwon97 committed:
feat(LineDao): 지하철 노선 삭제 기능 구현

Seongwon97 committed:
feat(LineController): 지하철 노선 저장 API 구현

Seongwon97 committed:
feat(LineController): 지하철 노선 목록 조회 API 구현

Seongwon97 committed:
feat(LineController): 지하철 노선 조회 API 구현

Seongwon97 committed:
feat(LineController): 지하철 노선 수정 API 구현

Seongwon97 committed:
feat(LineController): 지하철 노선 삭제 API 구현

Seongwon97 committed:
feat(LineController): 중복된 이름의 노선 등록시 예외 처리

Seongwon97 committed:
style: 코드 컨벤션 통일 & 네이밍 수정

Seongwon97 committed:
chore: h2 db 연결 설정

Seongwon97 committed:
refactor: spring IoC 컨테이너 사용

Seongwon97 committed:
chore: 테스트 db 환경 분리

Seongwon97 committed:
feat(StationDao): DB 연결 적용

Seongwon97 committed:
feat(LineDao): DB 연결 적용

Seongwon97 committed:
style: 불필요한 import문 제거 & 컨벤션 통일

Seongwon97 committed:
docs(README): 마크다운 출력 오류 수정

Seongwon97 commented on src/main/java/wooteco/subway/exception/DuplicateNameException.java:
```diff
@@ -0,0 +1,8 @@
+    public DuplicateNameException(String message) {
+        super(message);
+    }
+}
```
## 커스텀 예외를 만들 때, 상속 받는 예외 클래스의 선택
이번 미션을 진행하며 중복된 이름에 대한 예외처리와 DB로 부터 알맞는 값을 찾지 못하는 경우에는 커스텀 예외를 발생시키도록 코드를 작성했습니다. 그런데 여기서 커스텀 예외를 생성할 때 어떤 예외 클래스를 상속받는 것이 더 좋을지에 대하여 의문이 들더라고요.

상속 예외 클래스의 경우 `Exception`으로부터 `RuntimeException`, `IllegallArgumetException`까지 많은 클래스들이 존재하는데 보통 커스텀 예외 클래스의 경우, 어느정도 레벨의 예외를 상속하는게 좋을까요?

개인적인 생각으로는 중복된 이름의 오류의 경우, CheckedExeption인 `Exception`은 올바르지 못한 후보인 것 같고 UncheckedException인 `RuntimeException`, `IllegallArgumetException` 중에서는 어떤 클래스를 상속하던간에 `@ControllerAdvice`에서 예외를 잡아주기에 상관이 없다고 생각되면서도 그래도 가장 구체적인 하위의 예외를 상속하는게 더 좋을까?하는 생각도 드는 것 같습니다😭

Seongwon97 commented on src/main/java/wooteco/subway/controller/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,18 @@
+    public ResponseEntity<ErrorMessageResponse> duplicateStationNameException(DuplicateNameException e) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(new ErrorMessageResponse(e.getMessage()));
+    }
```
## 올바른 상태코드
지난 체스미션 때부터 지금까지 각각 예외 상황에 따른 반환 상태코드를 어떤 것을 선택하는게 올바른지 잘 모르겠습니다..😭

상태 코드를 정할 때마다 [Http Status](https://developer.mozilla.org/ko/docs/Web/HTTP/Status) 를 참조하는데도 실패에 따른 반환 상태코드 선택이 너무 어렵네요

현재 중복된 이름으로 데이터를 생성하는 Post Request를 보냈을 때, 발생하는 예외의 경우 400번대 코드를 반환해줬으나, 올바른 상태코드를 반환한 것인지 확신이 들지 않습니다.


- input값으로 서비스 로직 수행중 오류 발생하는 상황
- sql문 수행 중 예외 발생하는 상황
위의 상황의 경우 각각 어떤 상태코드를 반환해주는 것이 좋을까요😢

Seongwon97 commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,69 @@
+        };
+    }
+
+    public Line findById(Long id) {
+        String SQL = "select * from line where id = ?;";
+        try {
+            return jdbcTemplate.queryForObject(SQL, rowMapper(), id);
+        } catch (DataAccessException e) {
+            throw new NotFoundException("id에 맞는 지하철 노선이 없습니다.");
+        }
+    }
+
+    public void update(Line line) {
+        findById(line.getId());
```
## 존재하지 않는 데이터 Delete, Update의 예외처리
현재 update, delete의 경우, 요청이 왔을 때 해당 데이터들이 존재하는지 먼저 검증을 하기 위해 쿼리문 실행 전에 다음과 같이 `findById()` 메서드를 추가했습니다.

원래는 아래의 코드와 같이 예외처리를 하고 싶었으나, select문과는 다르게 delete의 경우 원하는 값이 존재하지 않을 경우 `EmptyResultDataAccessException`를 발생시키지 않아서 위와 같이 작성하였습니다.
```java
    public void delete(Long id) {
        String SQL = "delete from line where id = ?";
        try {
            jdbcTemplate.update(SQL, id);
        } catch (DataAccessException e) {
            throw new NotFoundException("id에 맞는 지하철 노선이 없습니다.");
        }
    }
```
코드들을 작성하며 존재하지 않는 값에 대한 delete, update 예외처리 방법들에 대해 생각해봤습니다.
1. 작성 코드와 같이 `findById()`를 통해 검증을 한다. -> 삭제나 업데이트를 하기 위해서 테이블 접근을 2번이나 해야한다.
2. 아무런 검증을 하지 않는다? ->  update와 delete의 메서드는 존재하지 않는 데이터에 대한 요청에도 `DataAccessException`을 발생시키지 않아, DB에 변경이 없어도 요청이 성공했다고 응답하게 된다.

개인적으로는 안정적인 서비스와 올바른 Response를 생각했을 때에는 테이블에 2번 접근하더라도 1번 방법이 더 좋을 것 같다고 생각합니다. 이러한 내용들을 과거 서비스에서는 어떤 방식으로 검증을 하는지 여쭙고 싶습니다.

Seongwon97 commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,69 @@
+        };
+    }
+
+    public Line findById(Long id) {
+        String SQL = "select * from line where id = ?;";
+        try {
+            return jdbcTemplate.queryForObject(SQL, rowMapper(), id);
+        } catch (DataAccessException e) {
+            throw new NotFoundException("id에 맞는 지하철 노선이 없습니다.");
+        }
```
## Query 실행할 때 발생하는 예외 처리의 위치 (DAO vs Service)
SQL문 실행에 대한 예외처리의 위치가 DAO에 있는 것이 좋을지, 아니면 Service에서 예외를 처리해주는 것이 좋을지에 대해서도 많은 고민을 했습니다.

현재는 DAO에서 예외처리를 추가해줬으나, 실제 서비스들에서는 jpa나 jdbc의 예외들을 어느 layer에서 처리를 하는지 궁금합니다🧐

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,54 @@
+
+    public void setColor(String color) {
+        this.color = color;
+    }
```
사용하고 있나요?

Seongwon97 replied:
사용하지 않는 Line클래스의 setter들을 모두 삭제하였습니다!
또한 Station의 기본 생성자도 사용되지 않아서 삭제했습니다.

choihz commented on src/main/java/wooteco/subway/controller/StationController.java:
```diff
@@ -1,39 +1,39 @@
     public ResponseEntity<List<StationResponse>> showStations() {
-        List<Station> stations = StationDao.findAll();
-        List<StationResponse> stationResponses = stations.stream()
-                .map(it -> new StationResponse(it.getId(), it.getName()))
-                .collect(Collectors.toList());
+        List<StationResponse> stationResponses = stationService.findAll();
         return ResponseEntity.ok().body(stationResponses);
     }
```
기본 제공된 코드이긴 하지만, 혹시 모르고 있었다면 이번 기회에 학습해 보아요. `produces` 설정은 어떤 역할을 할까요? 그리고 지금 이 설정을 지우면 어떻게 될까요?

Seongwon97 replied:
consumes와 produces는 이전 [next-step spring학습 레포지토리](https://github.com/next-step/spring-learning-test) 의 mvc branch에서 학습을 해본적이 있었습니다.

그때 학습을 하며 이해한 내용과 이번에 추가로 학습한 내용을 정리해보자면 두 설정은 request header의 값을 확인하며 controller로의 mapping을 제한해주는 것으로 이해했습니다. 이러한 속성 설정을 해둔다면 controller에서 보내는 데이터와 받는 데이터를 제한하여 오류 상황을 줄일 수 있는 것 같습니다.

- `@Consumes` : 수신 하고자하는 데이터 포맷을 정의한다.
- `@Produces `: 출력하고자 하는 데이터 포맷을 정의한다.

<hr>

해당 내용을 학습하며 [미션 API 문서](https://techcourse-storage.s3.ap-northeast-2.amazonaws.com/d5c93e187919493da3280be44de0f17f#_%EC%A7%80%ED%95%98%EC%B2%A0_%EB%85%B8%EC%84%A0_%EB%AA%A9%EB%A1%9D_http_response) 의 request와 response에 맞게 모든 controller 메서드에 consumes과 produces 속성을 아래와 같이 설정하고 몇가지 테스트를 해봤습니다. 
```java
    @GetMapping(value = "/stations",
            consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<List<StationResponse>> showStations() {
        List<StationResponse> stationResponses = stationService.findAll();
        return ResponseEntity.ok().body(stationResponses);
    }
```
위와 같이 지정을 하고 몇가지 테스트를 해봤는데 몇가지 의문이 생겨 질문을 남깁니다😭
1. `produces = MediaType.APPLICATION_JSON_VALUE`의 설정을 하였을 때, header에 `Accept: application/json` 이 없어도 controller에서는 해당 request를 받아서 처리해주더군요..accept값이 `application/json`타입이 아닌 다른 값으로 request를 보낼 경우 406과 같은 에러가 반환되지만 값이 없을 때는 왜 응답을 해주는지 궁급합니다. 제가 이해한 바로는 consumes, produces의 경우 header의 값을 확인하여 해당 값 설정이 존재하고 일치하여야지만 요청을 받아주는 것으로 이해를 하였는데 잘못 이해한 것일까요?? 아니면 Accept의 default값이 `application/json`인 것일까요..?? (consumes는 produces와 동일한 테스트를 해본 결과 해당 header값이 존재하지 않으면 415 Unsupported Media Type에러를 던져주는 것을 확인했습니다.)
2. consume, produces는 request나 response에 body값이 없더라도 api문서에 accept와 contextType이 지정되어 있으면 정의를 해주는 것이 좋은지 여쭙고 싶습니다. 해당 속성 설정을 해두면 데이터를 제한함으로써 오류 상황을 줄일 수 있다지만, body가 없는 요청,응답에 대해서는 해당 설정을 해주는 것이 과연 좋은지? 궁금합니다.

choihz commented on src/main/java/wooteco/subway/controller/StationController.java:
```diff
@@ -1,39 +1,39 @@
     public ResponseEntity<StationResponse> createStation(@RequestBody StationRequest stationRequest) {
-        Station station = new Station(stationRequest.getName());
-        Station newStation = StationDao.save(station);
-        StationResponse stationResponse = new StationResponse(newStation.getId(), newStation.getName());
-        return ResponseEntity.created(URI.create("/stations/" + newStation.getId())).body(stationResponse);
+        StationResponse stationResponse = stationService.create(stationRequest.getName());
+        return ResponseEntity.created(URI.create("/stations/" + stationResponse.getId())).body(stationResponse);
     }
```
요청 객체의 `name`에 빈 문자열을 넣어도 역이 생성되고 있네요 😭 또한, 255자가 넘는 값을 입력했을 때는 어떤 일이 일어나고 있나요?

Seongwon97 replied:
문자가 255가 넘는 값을 입력했을 때, 발생하는 예외에 대해 확실하게 알지 못하여서 테이블에서 name 문자열 크기 제한을 조정하여 테스트를 해본 결과 `DataIntegrityViolationException` 예외가 발생하는 것을 확인했습니다!!

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/57744251/167097046-0775bb02-0c3f-476b-97cd-caa090ba8082.png">

데이터를 넣어주기 전에 데이터 사이즈 검증을 추가하던가, 범위를 초과할 때 발생하는 예외를 처리해주는 작업을 추가해줘야 할 것 같아 `ExceptionHandler`를 통해 예외 처리를 해줬습니다!

choihz commented on src/main/java/wooteco/subway/controller/LineController.java:
```diff
@@ -0,0 +1,50 @@
+        return ResponseEntity.created(URI.create("/lines/" + lineResponse.getId())).body(lineResponse);
+    }
+
+    @GetMapping("/lines")
+    public ResponseEntity<List<LineResponse>> showLines() {
+        List<LineResponse> lineResponses = lineService.findAll();
+        return ResponseEntity.ok().body(lineResponses);
+    }
```
응답 형태가 [API 문서](https://techcourse-storage.s3.ap-northeast-2.amazonaws.com/d5c93e187919493da3280be44de0f17f#_%EC%A7%80%ED%95%98%EC%B2%A0_%EB%85%B8%EC%84%A0_%EB%AA%A9%EB%A1%9D_http_response)와는 다르네요.

```json
[ {
  "id" : 1,
  "name" : "신분당선",
  "color" : "bg-red-600"
}, {
  "id" : 2,
  "name" : "분당선",
  "color" : "bg-green-600"
} ]
```

Seongwon97 replied:
기본 제공 코드를 그대로 사용하다보니 `stations:null`데이터들이 모두 들어갔었네요..!
Response클래스를 수정해서 API 문서의 스펙과 동일하게 나오도록 수정하였습니다😁

choihz commented on src/main/java/wooteco/subway/controller/LineController.java:
```diff
@@ -0,0 +1,50 @@
+    @GetMapping("/lines")
+    public ResponseEntity<List<LineResponse>> showLines() {
+        List<LineResponse> lineResponses = lineService.findAll();
+        return ResponseEntity.ok().body(lineResponses);
```
```suggestion
        return ResponseEntity.ok(lineResponses);
```

사소한 것이긴 하지만, OK 응답의 경우 이렇게 shortcut을 지원하고 있어요. 또는, `ResponseEntity`로 감싸지 않고 응답 객체를 그대로 리턴해도 됩니다.

Seongwon97 replied:
지난 체스 미션을 스프링 프레임워크로 적용하면서 해당 내용을 분명 학습했는데..코드를 작성하다보니 다시 `body()`에 데이터를 넣어줬었네요😢

해당 내용도 수정했습니다!!

choihz commented on src/main/java/wooteco/subway/controller/LineController.java:
```diff
@@ -0,0 +1,50 @@
+        return ResponseEntity.ok().body(lineResponses);
+    }
+
+    @GetMapping("/lines/{lineId}")
+    public ResponseEntity<LineResponse> showLine(@PathVariable Long lineId) {
+        LineResponse lineResponse = lineService.findById(lineId);
+        return ResponseEntity.ok().body(lineResponse);
+    }
```
존재하지 않는 `lineId`로 조회하려고 하면 어떤 응답을 주고 있나요?

Seongwon97 replied:
404 Not Found 응답을 던져주는 것이 올바른데 실제로는 500 에러를 던져주고 있었네요..! 

수정했습니다!!😁

![image](https://user-images.githubusercontent.com/57744251/167133905-c513cab5-1380-4345-b166-7260f9dad7e6.png) 

`ExceptoinHandler`에서 잡아주는 에러는 기존 커스텀 예외를 제거하고, 찾는 값이 없을 때 발생하는 `EmptyResultDataAccessException`을 사용하도록 했습니다!

choihz commented on src/main/java/wooteco/subway/controller/LineController.java:
```diff
@@ -0,0 +1,50 @@
+        this.lineService = lineService;
+    }
+
+    @PostMapping("/lines")
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
+        LineResponse lineResponse = lineService.create(lineRequest.getName(), lineRequest.getColor());
+        return ResponseEntity.created(URI.create("/lines/" + lineResponse.getId())).body(lineResponse);
+    }
```
`LineService`에 값을 하나하나 꺼내서 전달하고 있는데요, 만약 전달해야 할 값이 10개쯤 된다면 어떻게 될까요? 😢

Seongwon97 replied:
해당 피드백은 아래의 피드백과 함께 고민을 해봤습니다.

고민을 해본 결과는 다음과 같습니다.
![IMG_333D451681C7-1](https://user-images.githubusercontent.com/57744251/167152732-1b4546dc-046a-44e8-867a-5812e81d2906.jpeg)

controller에서 service로 데이터를 넘겨줄 때, RequestBody로부터 값을 하나씩 뽑아서 전달할 경우 코드의 복잡도가 증가하고, 보내주는 데이터의 스펙이 변경되었을 때 컨트롤러로부터 서비스, dto까지 모든 코드를 수정해야한다는 문제점이 있다고 느꼈습니다.

그래서 해당 문제는 컨트롤러에서 서비스로 넘겨줄 때는 request dto를 넘겨주어 필요한 값을 서비스에서 추출하여 사용하도록 수정하였습니다.

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,54 @@
+        Line line = new Line(name, color);
+        validateDuplicationName(line);
+        Line savedLine = lineDao.save(line);
+        return new LineResponse(savedLine.getId(), savedLine.getName(), savedLine.getColor(), null);
```
`LineResponse`를 생성하는 과정에서도 값을 하나하나 꺼내 전달하고 있어요. `Line`의 정보가 수정되면 `LineResponse`를 생성하는 모든 코드를 찾아 함께 수정해야 하지 않을까요? 더 나은 방법은 없을까요?

Seongwon97 replied:
코니의 말대로 Response DTO를 생성할 때 기존 코드와 같이 도메인 객체로부터 getter메서드를 활용해 모든 내부 값들을 추출하여 DTO를 생성할 경우, 도메인의 변경이 있을 경우 많은 서비스 로직을 변경해야한다는 문제가 있는 것 같습니다ㅠ

해당 문제를 해결하려고 도메인 객체를 받아 Response객체를 반환하는 정적 팩터리 메서드를 생성하여 사용하도록 수정했습니다. 😁

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,54 @@
+        return new LineResponse(savedLine.getId(), savedLine.getName(), savedLine.getColor(), null);
+    }
+
+    private void validateDuplicationName(Line line) {
+        List<Line> lines = lineDao.findAll();
+        if (lines.contains(line)) {
+            throw new DuplicateNameException("중복된 지하철 노선 이름이 있습니다.");
+        }
+    }
```
이름 중복 확인을 위해 모든 데이터를 가져와야만 할까요? 데이터가 엄청 많다면 어떻게 될까요?

Seongwon97 replied:
![image](https://user-images.githubusercontent.com/57744251/167135917-2f2309d8-9b2e-4d08-b05b-fa7c6d280eeb.png)

해당 내용도 데이터 중복 체크를 위해 `findAll()`을 통해 모든 데이터를 불러오는 대신 테이블을 생성한 DDL에서 name column에 unique를 붙여준 것을 활용하여 중복된 데이터가 저장될 때 발생하는 `DuplicateKeyException`를 활용하도록 수정하였습니다.😁

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,54 @@
+        if (o == null || getClass() != o.getClass()) return false;
+        Line line = (Line) o;
+        return getName().equals(line.getName()) && getColor().equals(line.getColor());
+    }
```
`name`과 `color`가 같으면 같은 객체로 보는군요! 그렇다면 `Line`에서 `id`는 어떤 의미를 갖나요?

Seongwon97 replied:
피드백을 받고 나서 코드를 전체적으로 보니 기존의 equals와 hashCode는 test를 할 때 제 편의를 위한 재정의였던 것 같네요..😅

해당 내용을 id까지 같아야지 같은 객체로 보도록 수정하고 test코드들도 그에 맞게 수정하였습니다.

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,54 @@
+    private String name;
+    private String color;
+
+    public Line(String name, String color) {
+        this.name = name;
+        this.color = color;
+    }
+
+    public Line(Long id, String name, String color) {
+        this.id = id;
+        this.name = name;
+        this.color = color;
+    }
```
자바에서는 하나의 생성자에서 다른 생성자를 호출할 수 있어요. 이 방법을 사용해 볼까요? 어떤 장점이 있을까요?

Seongwon97 replied:
피드백을 반영하여 점층적 생성자 패턴을 적용하였습니다😁

choihz commented on src/main/java/wooteco/subway/controller/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,18 @@
+    public ResponseEntity<ErrorMessageResponse> duplicateStationNameException(DuplicateNameException e) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(new ErrorMessageResponse(e.getMessage()));
+    }
```
`DuplicateNameException`만 처리해 주어도 괜찮을까요? 다른 종류의 예외가 발생하면 어떻게 될까요?

choihz replied:
> 지난 체스미션 때부터 지금까지 각각 예외 상황에 따른 반환 상태코드를 어떤 것을 선택하는게 올바른지 잘 모르겠습니다..😭
> 
> 상태 코드를 정할 때마다 [Http Status](https://developer.mozilla.org/ko/docs/Web/HTTP/Status) 를 참조하는데도 실패에 따른 반환 상태코드 선택이 너무 어렵네요
> 
> 현재 중복된 이름으로 데이터를 생성하는 Post Request를 보냈을 때, 발생하는 예외의 경우 400번대 코드를 반환해줬으나, 올바른 상태코드를 반환한 것인지 확신이 들지 않습니다.
> 
> - input값으로 서비스 로직 수행중 오류 발생하는 상황
> - sql문 수행 중 예외 발생하는 상황
> 위의 상황의 경우 각각 어떤 상태코드를 반환해주는 것이 좋을까요😢

이 상황에서 400은 충분히 적절해 보이는데 어떤 부분에서 불안함을 느끼시는 걸까요?

알려주신 링크에서도 볼 수 있듯 굉장히 많은 상태코드가 존재하지만, 사실 자주 사용하는 건 꽤나 한정적인 것 같아요. 그래서 경험이 조금 더 쌓이면 많은 문제가 해결될 것이라고 생각해요. 그리고 완벽하게 딱 들어맞는 코드를 찾겠다는 집착 같은 건 하지 않아도 됩니다 🙂

더불어, 제시하신 상황에 적절한 코드는 지금 코멘트로 남겨 주신 맥락만 보고는 판단하기 어려울 것 같아요. 예를 들어 같은 "sql문 수행 중 예외 발생하는 상황"이라고 해도 실제 (비즈니스적 맥락에서의?) 원인은 굉장히 다양할 수 있지 않을까요?

Seongwon97 replied:
> 이 상황에서 400은 충분히 적절해 보이는데 어떤 부분에서 불안함을 느끼시는 걸까요?

사실 불안감보다는 스프링 개발 경험이 적어 상황별 상태코드 사용에 대한 자신감이 없던 것 같습니다. 코니의 말대로 경험이 개발 경험을 쌓이면 이러한 고민들도 사라질 거라 믿습니다😁

<hr>

> 더불어, 제시하신 상황에 적절한 코드는 지금 코멘트로 남겨 주신 맥락만 보고는 판단하기 어려울 것 같아요. 예를 들어 같은 "sql문 수행 중 예외 발생하는 상황"이라고 해도 실제 (비즈니스적 맥락에서의?) 원인은 굉장히 다양할 수 있지 않을까요?

질문을 올릴때까지는 SQL문을 실행할 때 발생하는 예외가`DataAccessException`으로만 발생하는줄 알고 "SQL문 에러는 하나의 상태코드로 반환해야한다!"라는 생각을 갖고 있었습니다.
하지만 코니의 피드백들을 반영해가며 많은 디버깅들을 하다보니 SQL문의 수행도 각각 상황에 따라 다른 예외가 발생하는 것을 확인했네요😅  

SQL수행 예외만 봐도 다양한 원인이 존재하는데 각각 상황에 대한 대표 예외 코드에 대한 질문은 잘못된 질문이었네요..!😭

이러한 질문에도 친절히 답해주셔서 감사합니다🙇🏻‍♂️

Seongwon97 replied:
> DuplicateNameException만 처리해 주어도 괜찮을까요? 다른 종류의 예외가 발생하면 어떻게 될까요?

피드백을 수정하며 [중복 값 체크 로직 관련 피드백](https://github.com/woowacourse/atdd-subway-map/pull/229#discussion_r866158012) 와 같이 커스텀 예외를 삭제하고 중복 값은 `DuplicateKeyException`예외를 잡아주도록 수정하였습니다. 
그리고 혹여 handler를 통해 잡지 못한 sql관련 오류가 존재할 수 있기에 `DuplicateKeyException`에 대한 예외 처리도 추가했습니다😁

choihz commented on src/main/java/wooteco/subway/controller/LineController.java:
```diff
@@ -0,0 +1,50 @@
+    public ResponseEntity<Void> updateLine(@PathVariable Long lineId, @RequestBody LineRequest lineRequest) {
+        lineService.update(lineId, lineRequest.getName(), lineRequest.getColor());
+        return ResponseEntity.ok().build();
+    }
```
다른 `Line`이 이미 가지고 있는 이름으로 수정하려고 할 때, 어떤 응답을 주고 있나요?

Seongwon97 replied:
기존 로직으로는 업데이트시에 이름 중복 체크 관련 로직과 sql문 실행 오류에 따른 exceptionHandler도 없어서 500에러가 발생하였을 것 같습니다.

지금은 `DuplicateKeyException`을 handler에서 잡아주어 `400 Bad Request`에 중복된 데이터가 존재하다는 메시지를 반환해주고 있습니다.

choihz commented on src/main/java/wooteco/subway/exception/DuplicateNameException.java:
```diff
@@ -0,0 +1,8 @@
+    public DuplicateNameException(String message) {
+        super(message);
+    }
+}
```
> 상속 예외 클래스의 경우 `Exception`으로부터 `RuntimeException`, `IllegallArgumetException`까지 많은 클래스들이 존재하는데 보통 커스텀 예외 클래스의 경우, 어느정도 레벨의 예외를 상속하는게 좋을까요?
> 
> 개인적인 생각으로는 중복된 이름의 오류의 경우, CheckedExeption인 `Exception`은 올바르지 못한 후보인 것 같고 UncheckedException인 `RuntimeException`, `IllegallArgumetException` 중에서는 어떤 클래스를 상속하던간에 `@ControllerAdvice`에서 예외를 잡아주기에 상관이 없다고 생각되면서도 그래도 가장 구체적인 하위의 예외를 상속하는게 더 좋을까?하는 생각도 드는 것 같습니다😭

당연히 이런 문제에는 정답이 없겠죠?

그런데, `DuplicateNameException`이 `RuntimeException`을 상속할 때와 `IllegallArgumetException`을 상속할 때 `@ControllerAdvice`에서 예외를 처리하는 방식에 큰 차이가 없을까요? 지하철 노선도 프로그램의 요구사항이 점점 커지고 다양한 예외가 발생할 수 있는 상황을 가정해 보았을 때, 만약 커스텀 예외가 전부 `RuntimeException`을 상속했다면 어떤 문제가 발생할 수 있을지 고민해 보면 좋을 것 같아요.

Seongwon97 replied:
커스텀 예외 클래스가 `RuntimeException`나 `IllegallArgumetException`를 상속받을 때 모두 커스텀 예외 클래스의 이름으로 `@ControllerAdvice`에서 잡아줘서 발생할 수 있을 문제가 떠오르지 않네요.. 

머리속으로는 더욱더 세분화된 `IllegallArgumetException`와 같은 자식클래스를 상속 받는 것이 더 좋을거라 생각되지만 이유를 설명하지 못하겠네요😅

혹시 관련하여 학습할 수 있는 키워드가 있을까요..?

choihz commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,69 @@
+        } catch (DataAccessException e) {
+            throw new NotFoundException("id에 맞는 지하철 노선이 없습니다.");
+        }
+    }
```
> SQL문 실행에 대한 예외처리의 위치가 DAO에 있는 것이 좋을지, 아니면 Service에서 예외를 처리해주는 것이 좋을지에 대해서도 많은 고민을 했습니다.
> 
> 현재는 DAO에서 예외처리를 추가해줬으나, 실제 서비스들에서는 jpa나 jdbc의 예외들을 어느 layer에서 처리를 하는지 궁금합니다🧐

JPA를 사용해 보셨는지 모르겠으나 지금 우리의 DAO를 이용하는 방식과는 다소 달라서, 동일 선상에 놓고 비교하기는 어려울 것 같아요.

그럼 고민해 볼까요? DAO의 역할은 어디까지여야 할까요?

Seongwon97 replied:
피드백을 반영하며 SQL문의 실행에서 발생하는 예외의 종류가 많다는 것을 알게 되었습니다.

지금 드는 생각으로는 DAO에서는 DB와의 통신만 수행하고, 쿼리문 실행중 발생하는 예외들은 Handler에서 잡아서 처리를 해주는 것이 좋은 방법이라 생각됩니다.

쿼리문 실행시 발생하는 예외 외에 값이 유효한 값인지에 대한 테스트는 서비스 로직 또는 도메인 로직에 위치하는 것이 좋다고 생각됩니다😎

choihz commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+
+    public void update(Line line) {
+        findById(line.getId());
+        String SQL = "update line set name = ?, color = ? where id = ?;";
+        jdbcTemplate.update(SQL, line.getName(), line.getColor(), line.getId());
+    }
```
> 코드들을 작성하며 존재하지 않는 값에 대한 delete, update 예외처리 방법들에 대해 생각해봤습니다.
> 1. 작성 코드와 같이 `findById()`를 통해 검증을 한다. -> 삭제나 업데이트를 하기 위해서 테이블 접근을 2번이나 해야한다.
> 2. 아무런 검증을 하지 않는다? ->  update와 delete의 메서드는 존재하지 않는 데이터에 대한 요청에도 `DataAccessException`을 발생시키지 않아, DB에 변경이 없어도 요청이 성공했다고 응답하게 된다.
> 
> 개인적으로는 안정적인 서비스와 올바른 Response를 생각했을 때에는 테이블에 2번 접근하더라도 1번 방법이 더 좋을 것 같다고 생각합니다. 이러한 내용들을 과거 서비스에서는 어떤 방식으로 검증을 하는지 여쭙고 싶습니다.

당연히 2번은 옳은 방향이 아니라고 생각해요. 존재하지 않는 식별자로 요청을 보낸다는 것은 분명히 문제가 있는 상황이기 때문에 사용자에게 분명히 알리고 서버에서도 알아야 하겠죠.

그러나, 위에서도 코멘트를 남긴 것처럼 검증을 어디서 해야 하는지는 고민해 봐야 할 것 같아요. 그리고, `findById()` 메서드를 바로 호출하는 방법이 최선일까요? 이 방식으로 코드의 의도가 분명히 드러나나요?

Seongwon97 replied:
해당 방식에 의도가 잘 드러나지 않는 것 같아 dao에 해당 id를 가진 데이터가 존재하는지 체크하는 `existLineById()`메서드를 만들어 데이터 존재 유무를 확인하도록 수정했습니다😎

choihz commented on src/test/java/wooteco/subway/dao/LineDaoTest.java:
```diff
@@ -0,0 +1,94 @@
+        assertThat(actual).isEqualTo(expected);
+    }
+
+    @DisplayName("노선을 삭제한다.")
+    @Test
+    void delete() {
+        Line saveLine = lineDao.save(line);
+
+        lineDao.delete(saveLine.getId());
+
+        assertThatThrownBy(() -> lineDao.findById(saveLine.getId()))
+                .isInstanceOf(NotFoundException.class)
+                .hasMessageMatching("id에 맞는 지하철 노선이 없습니다.");
+    }
```
테스트 이름이 테스트의 내용을 잘 설명하지 못하고 있는 것 같아요 🤔

Seongwon97 replied:
테스트 이름에 맞도록 해당 내용을 수정했습니다!!

처음에는 데이터가 잘 삭제되었다는 것을 어떻게 테스트 할지 몰라서 위와 같이 작성하였는데, `assertThat(lines.contains(savedLine)).isFalse();`와 같이 해당 데이터의 존재 여부를 검증하면 됐었네요..ㅎ

choihz commented on src/test/java/wooteco/subway/controller/LineControllerTest.java:
```diff
@@ -0,0 +1,143 @@
+    @DisplayName("지하철 노선을 등록한다.")
+    @Test
+    void createLine() {
+        Map<String, String> params = new HashMap<>();
+        params.put("name", "신분당선");
+        params.put("color", "red");
```
`Map` 대신 `LineRequest`를 이용해도 되지 않을까요?

Seongwon97 replied:
앗..그렇네요!! 
코니의 말대로 `Map`대신 `LineRequest`를 사용하도록 수정했습니다!!

choihz commented on src/test/java/wooteco/subway/controller/LineControllerTest.java:
```diff
@@ -0,0 +1,143 @@
+                .extract();
+        List<LineResponse> actual = response.jsonPath().getList(".", LineResponse.class);
+
+        assertThat(actual.size()).isEqualTo(2);
```
```suggestion
        assertThat(actual).hasSize(2);
```

이런 메서드도 있어요 😎

그런데, 크기 검증만으로 충분할까요? 내부에서 값을 아무렇게나 바꿔서 리턴해도 크기만 같으면 테스트가 통과하게 되는 것 아닐까요?

Seongwon97 replied:
`assertThat().hasSize()`메서드는 처음 본 것 같네요🥸!!

반환 리스트의 크기만 검증하는 테스트 코드들을 내부 값들을 검증하도록 수정했습니다😁

choihz reviewed (changes requested):
안녕하세요, 렉스! 다시 만나게 되어 반가워요 👋

지하철 노선도 미션의 1, 2단계를 잘 구현해 주셨네요 👍 피드백을 남기며 주신 질문에 대한 답도 달아 두었습니다. 궁금한 점이나 의견이 있으시다면 언제나 코멘트 또는 DM 주세요!

Seongwon97 committed:
refactor: 불필요한 메서드 제거

Seongwon97 committed:
feat: 잘못된 사이즈의 데이터로 데이터를 생성하는 것에 대한 예외 처리 추가

Seongwon97 committed:
refactor(LineResponse): API스펙에 맞게 응답 형태 수정

Seongwon97 committed:
refactor(Controller): ResponseEntity.ok의 shortcut기능 적용

ResponseEntity.ok의 shortcut기능을 적용해 body의 값을 바로 넣어주도록 수정

Seongwon97 committed:
refactor: db에 없는 데이터를 조회하는 요청이 올 떄, 404상태코드를 반환하도록 수정

- 커스텀 Exception인 NotFoundException을 제거하고, ControllerAdvice에서 ExceptionHanlerController를 잡도록 설정

Seongwon97 committed:
refactor: 데이터 중복 체크 로직 수정

- 모든 데이터를 불러와 service에서 체크하는 기존 로직을 중복 데이터 저장시 발생하는 DuplicateKeyException을 처리하는 방법으로 수정

Seongwon97 committed:
refactor: 도메인에 점층적 생성자 패턴 적용 & final 키워드 추가

Seongwon97 committed:
refactor: 컨트롤러와 서비스 계층 간 데이터 전달 로직 수정

- 컨트롤러에서 서비스 계층으로 데이터를 전달할 때는 DTO를 통해 전달하도록 수정
- 서비스에서 컨트롤러로 전달하는 DTO를 정적 팩터리 메서드로 도메인 객체를 받아서 생성하도록 수정

Seongwon97 committed:
refactor: 도메인 객체의 equals & hashCode에 id를 포함하도록 수정

Seongwon97 committed:
refactor(ExceptionHandlerController): 메서드 이름 수정 및 DataAccessException예외 처리 추가

Seongwon97 committed:
refactor: 컨트롤러 테스트 요청 테이터에 Request클래스를 사용하도록 수정

Seongwon97 committed:
test: 노선 목록 조회 테스트 코드 수정

- size만 체크하던 테스트에서 내부 값을 비교하도록 수정

Seongwon97 committed:
test: 지하철 노선 삭제 테스트 이름에 맞도록 테스트 내용 수정

Seongwon97 committed:
refactor: 존재하지 않는 데이터에 대한 update,delete 예외처리 로직 수정

Seongwon97 commented:
안녕하세요 코니👋🏻

보내주신 피드백들을 반영해봤습니다!!
이번에도 코니의 리뷰 덕분에 많은 것을 배우게 된 것 같습니다!!

바쁘신 와중에도 시간내어 리뷰해주셔서 감사합니다🙇🏻‍♂️
이번에도 잘 부탁드립니다!!

choihz commented on src/main/java/wooteco/subway/controller/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,46 @@
+    @ExceptionHandler(EmptyResultDataAccessException.class)
+    public ResponseEntity<ErrorMessageResponse> handleEmptyResultDataAccessException() {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND)
+                .body(new ErrorMessageResponse("요청한 리소스를 DB에서 찾을 수 없습니다."));
```
"DB에서" 찾을 수 없다는 것까지 알려줄 필요가 있을까요? 물론 정답이 없겠지만, 예외 메시지는 어느 정도까지 자세한 것이 좋을까요?

Seongwon97 replied:
> "DB에서" 찾을 수 없다는 것까지 알려줄 필요가 있을까요? 

굳이 DB에서 찾을 수 없다!라는 메시지를 보낼 필요는 없는 것 같습니다. 
간단하게 "요청한 리소스를 찾을 수 없습니다."의 메시지만 출력해줘도 괜찮을 것 같습니다.

> 물론 정답이 없겠지만, 예외 메시지는 어느 정도까지 자세한 것이 좋을까요?

지금까지 미션을 진행하며 느낀바로는 예외 메시지를 상세하게 내보내주는 것과 코드를 간단하게 작성하는 것은 tradeoff한 관계라고 생각됩니다. 예외 메시지를 상세하게 담아서 exception을 발생시키기 위해서는 try/catch를 통해 예외를 잡고 각각 상황에 맞는 메시지를 담아서 보내야하는데, 이런 경우에는 코드의 길이가 길어지고 가독성이 떨어지는 문제도 있는 것 같다 생각됩니다.

그래서 현재 저는 서비스 로직과 관련있는 내용의 경우, 서비스 로직내에서 메시지를 담은 예외를 던져두고 조회 예외, 중복키 관련 예외와 같은 예외들은 exceptionHandler를 통한 통합 예외를 던져주도록 설정했습니다.

choihz commented on src/main/java/wooteco/subway/controller/ExceptionHandlerController.java:
```diff
@@ -0,0 +1,46 @@
+                .body(new ErrorMessageResponse("요청한 리소스를 DB에서 찾을 수 없습니다."));
+    }
+
+    @ExceptionHandler(DuplicateKeyException.class)
+    public ResponseEntity<ErrorMessageResponse> handleDuplicateKeyException() {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(new ErrorMessageResponse("중복된 데이터가 존재합니다."));
+    }
+
+    @ExceptionHandler(DataAccessException.class)
+    public ResponseEntity<ErrorMessageResponse> handleDataAccessException() {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(new ErrorMessageResponse("DB관련 작업에서 오류가 발생했습니다."));
+    }
```
지금 이 `ExceptionControllerAdvice`가 속한 패키지는 `ui`인데, 이 메서드들이 다루는 예외는 스프링의 `dao` 패키지에 속해 있네요. `dao`의 예외들은 어디서 처리하는 것이 좋을지 고민해 보아도 좋을 것 같아요.

choihz commented on src/test/java/wooteco/subway/dao/LineDaoTest.java:
```diff
@@ -0,0 +1,105 @@
+        assertThat(actual).isEqualTo(expected);
+    }
+
+    @DisplayName("id에 맞는 노선이 없을 경우 예외를 발생시킨다.")
+    @Test
+    void findByIdException() {
+        assertThatThrownBy(() -> lineDao.findById(1L))
+                .isInstanceOf(EmptyResultDataAccessException.class);
+    }
```
어떤 상황에서 id에 맞는 노선이 없을 때 예외가 발생하는지, 상황을 테스트 이름에 더 잘 드러내면 좋을 것 같아요.

Seongwon97 replied:
넵!! DisplayName에 "노선 조회중 id에 맞는 노선이 없을 경우 예외를 발생시킨다."로 수정하여 예외를 발생하는 상황을 나타내도록 수정했습니다!!

choihz commented on src/test/java/wooteco/subway/dao/LineDaoTest.java:
```diff
@@ -0,0 +1,105 @@
+                .isInstanceOf(EmptyResultDataAccessException.class);
+    }
+
+    @DisplayName("노선의 이름과 색깔을 수정한다.")
+    @Test
+    void update() {
+        Line saveLine = lineDao.save(line);
+        Line expected = new Line(saveLine.getId(), "다른 분당선", "green");
+
+        lineDao.update(expected);
+        Line actual = lineDao.findById(saveLine.getId());
+
+        assertThat(actual).isEqualTo(expected);
+    }
+
+    @DisplayName("노선을 삭제한다.")
+    @Test
+    void delete() {
+        Line saveLine = lineDao.save(line);
+
+        lineDao.delete(saveLine.getId());
+        List<Line> lines = lineDao.findAll();
+
+        assertThat(lines.contains(saveLine)).isFalse();
+    }
```
수정과 삭제 작업에서도, happy path만 테스트하지 말고 예외적인 상황에서 예상한 대로 동작하는지 검증하는 테스트를 작성해 두면 좋지 않을까요? 존재하지 않는 리소스를 대상으로 한 요청, 컬럼 크기를 초과하는 데이터로의 수정 요청, 중복 이름으로의 수정 요청 등 다양한 상황이 존재하겠네요.

choihz commented on src/main/java/wooteco/subway/exception/AccessNoneDataException.java:
```diff
@@ -0,0 +1,7 @@
+package wooteco.subway.exception;
+
+public class AccessNoneDataException extends RuntimeException {
+
+    public AccessNoneDataException() {
+    }
+}
```
이 예외의 생성자는 `DataLengthException`의 생성자와 모양새가 다르군요. `ExceptionHandlerController`에서 항상 고정 메시지를 노출하기로 결정하셨기 때문일까요?

Seongwon97 replied:
처음에는 고정 메시지를 출력하기 위해 생성자가 메시지를 담지 않도록 하였는데, 위의 피드백을 반영하며 생각해보니 고정 예외 메시지를 내보낼 거면 해당 예외 클래스를 생성하지 않고 `EmptyResultDataAccessException`로 내보내면 되는데..왜 커스텀 예외를 만들었지?라는 생각이 들었습니다.

해당 `AccessNoneDataException`예외는 서비스 실행 도중 발생시킨 예외라 더욱 구체적인 상황을 담아 내보내면 예외 지점을 알아차리기 쉬울 것 같아 메시지를 담도록 수정했습니다.

choihz commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,54 @@
+
+    private void validateDataSize(String name) {
+        if (name.isEmpty() || name.length() > 255) {
+            throw new DataLengthException("역 이름이 빈 값이거나 최대 범위를 초과했습니다.");
```
1. 사용자에게 직접 노출되는 메시지라면, 최대 범위가 얼마인지 명시해 준다면 더 좋겠죠?
2. 이 역할을 서비스에서 해야 할까요? 더 좋은 방법은 없을까요?

Seongwon97 replied:
> 1. 사용자에게 직접 노출되는 메시지라면, 최대 범위가 얼마인지 명시해 준다면 더 좋겠죠?

더욱 구체적인 내용을 담도록 수정했습니다!

> 2. 이 역할을 서비스에서 해야 할까요? 더 좋은 방법은 없을까요?

해당 로직은 리펙터링을 하며 도메인으로 이동시켰습니다!!

choihz commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private void validateExistData(Long lineId) {
+        boolean isExist = stationDao.existStationById(lineId);
```
[Bool 변수 이름 제대로 짓기 위한 최소한의 영어 문법](https://soojin.ro/blog/naming-boolean-variables)을 한번 읽어 보면 좋을 것 같아요!

> ❗주의❗is로 시작하는 변수명을 짓다가 범하는 흔한 실수가 바로 is + 동사원형 을 쓰는 것이다.
> 
> isAuthorize, isHide, isFind 등등.

Seongwon97 replied:
is+형용사 문법을 이용하여 변수명을 수정하였습니다!!

좋은 자료 공유해주셔서 감사합니다🙇🏻‍♂️

choihz commented on src/main/java/wooteco/subway/exception/DataLengthException.java:
```diff
@@ -0,0 +1,8 @@
+package wooteco.subway.exception;
+
+public class DataLengthException extends RuntimeException {
+
+    public DataLengthException(String message) {
+        super(message);
+    }
+}
```
> 커스텀 예외 클래스가 `RuntimeException`나 `IllegallArgumetException`를 상속받을 때 모두 커스텀 예외 클래스의 이름으로 `@ControllerAdvice`에서 잡아줘서 발생할 수 있을 문제가 떠오르지 않네요.. 
> 
> 머리속으로는 더욱더 세분화된 `IllegallArgumetException`와 같은 자식클래스를 상속 받는 것이 더 좋을거라 생각되지만 이유를 설명하지 못하겠네요😅
> 
> 혹시 관련하여 학습할 수 있는 키워드가 있을까요..?

아시겠지만 `RuntimeException`은 굉장히 상위의 예외 클래스예요. 그래서 조금 더 구체적인 예외를 이용한다면 커스텀 예외를 어떤 상황을 위해 만든 것인지 조금 더 쉽게 파악할 수 있는 장점이 있을 것 같아요.

또한, 지금은 모든 커스텀 예외를 `ExceptionHandlerController`에서 일일이 처리하고 있지만, 만약 프로그램이 커지고 커스텀 예외가 100개쯤 된다면 이걸 일일이 다 처리할 수 있을까요? 적당한 계층 구조를 만들면 가장 아래에 있는 예외들은 굳이 `@ExceptionHandler`를 통해 처리하지 않고, 상위의 예외를 통해 일괄 처리할 수 있을 거예요.

혹시 잘 이해가 안 되었다면 `@ExceptionHandler`가 어떤 방식으로 예외를 처리하는지 학습해 보아요.

choihz commented on src/main/java/wooteco/subway/controller/StationController.java:
```diff
@@ -1,39 +1,39 @@
-        List<Station> stations = StationDao.findAll();
-        List<StationResponse> stationResponses = stations.stream()
-                .map(it -> new StationResponse(it.getId(), it.getName()))
-                .collect(Collectors.toList());
-        return ResponseEntity.ok().body(stationResponses);
+        List<StationResponse> stationResponses = stationService.findAll();
+        return ResponseEntity.ok(stationResponses);
     }
```
이 부분에서 남긴 코멘트에 대해, 정말 꼼꼼하게 학습하셨군요 👏

> 1. `produces = MediaType.APPLICATION_JSON_VALUE`의 설정을 하였을 때, header에 `Accept: application/json` 이 없어도 controller에서는 해당 request를 받아서 처리해주더군요..accept값이 `application/json`타입이 아닌 다른 값으로 request를 보낼 경우 406과 같은 에러가 반환되지만 값이 없을 때는 왜 응답을 해주는지 궁급합니다. 제가 이해한 바로는 consumes, produces의 경우 header의 값을 확인하여 해당 값 설정이 존재하고 일치하여야지만 요청을 받아주는 것으로 이해를 하였는데 잘못 이해한 것일까요?? 아니면 Accept의 default값이 `application/json`인 것일까요..?? (consumes는 produces와 동일한 테스트를 해본 결과 해당 header값이 존재하지 않으면 415 Unsupported Media Type에러를 던져주는 것을 확인했습니다.)

기회가 된다면, 우리가 HTTP 요청을 보냈을 때 스프링이 어떻게 해당 요청을 처리할 수 있는 컨트롤러의 메서드를 찾아내는지에 관해 학습해 보시면 좋을 것 같아요.

그리고, `Accept` 헤더의 기본값은 당연히 `application/json`이 아닙니다. ([참고](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2)) 다만, 요청하는 입장에서 받을 수 있는 타입을 제한하지 않았다면 당연히 보낼 수 있는 값을 그냥 보내주는 것이 합리적인 선택 아닐까요?

> 2. consume, produces는 request나 response에 body값이 없더라도 api문서에 accept와 contextType이 지정되어 있으면 정의를 해주는 것이 좋은지 여쭙고 싶습니다. 해당 속성 설정을 해두면 데이터를 제한함으로써 오류 상황을 줄일 수 있다지만, body가 없는 요청,응답에 대해서는 해당 설정을 해주는 것이 과연 좋은지? 궁금합니다.

~이건 프로젝트 컨벤션에 따라 다를 것 같아요. 팀에서 명시하기로 정했다면 지정해 줘야 하겠죠?~

~다만, 요즘은 대부분 json만을 이용하는 추세라서 굳이 명시하지 않고, 특별히 다른 타입을 사용하는 경우가 있다면 그때만 지정해 주는 것 같습니다.~

질문의 핵심을 잘못 이해했네요 😅
body가 없는 경우에 대해서는, 해당 헤더의 역할을 잘 이해하셨다면 설정할 필요가 있는지 스스로 판단이 가능할 것 같은데, 어떤가요?

choihz reviewed (approved):
안녕하세요, 렉스!

수정해 주신 내용을 확인하고 몇 가지 코멘트와 질문에 대한 답을 남겨 두었습니다. 1, 2단계 고생 많으셨고, 다음 단계에서 다시 만나요~ 👋
